include/seqan3/core/parallel/detail/reader_writer_manager.hpp

```diff
@@ -0,0 +1,315 @@
+            throw std::invalid_argument{"Both, reader count and writer count must be at least 1."};
+    }
+    //!\}
+
+    /*!\brief Atomically decrements writer counter by one and blocks the calling thread.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for producer threads and blocks the calling thread until all participating
+     * threads have arrived. If all participating producer threads arrived at the synchronisation point, one of the
+     * threads will invoke the completion function.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void writer_arrive_and_wait() noexcept
+    {
+        writer_latch.arrive_and_wait();
+
+        std::call_once(flag, completion_fn);
+    }
+
+    /*!\brief Atomically decrements writer counter by one.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for producer threads without blocking the calling thread.
+     * If all participating producer threads arrived at the synchronisation point, one of the
+     * threads will invoke the completion function.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void writer_arrive() noexcept
+    {
+        writer_latch.arrive();
+
+        if (writer_latch.try_wait())
+            std::call_once(flag, completion_fn);
+    }
+
+    /*!\brief Atomically decrements writer counter by one and blocks the calling thread.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for consumer threads and blocks the calling thread until all participating
+     * threads have arrived.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void reader_arrive_and_wait() noexcept
+    {
+        reader_latch.arrive_and_wait();
+    }
+
+    /*!\brief Atomically decrements reader counter by one.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for consumer threads without blocking the calling thread.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void reader_arrive() noexcept
+    {
+        reader_latch.arrive();
+    }
+
+    /*!\brief Registers the current thread as a producer thread for the monitored resource.
+     * \returns A RAII wrapper class that automatically deregisters from the monitored resource when destructed.
+     *
+     * \details
+     *
+     * If a thread calls this function it will become a participating producer thread for the monitored concurrent
+     * data structure. On destruction of the returned RAII wrapper class
+     * seqan3::detail::reader_writer_manager::writer_arrive is called automatically.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
```

Comment:
:nail_care: 

I think we write full sentences for this or at least "Sentence case."
```
     * ### Exception
     *
     *  Guaranteed not to throw.
     *
     * ### Thread safety
     *
     *  Thread-safe.
```

I think we should actually make doxygen macros for this, but that can happen after the release.